src/generators/net_caffe_recurrent.cc

```diff
@@ -0,0 +1,371 @@
+    lparam->add_top(cont_seq);
+
+    caffe::SliceParameter *sparam = lparam->mutable_slice_param();
+    sparam->set_slice_dim(2);
+    for (std::set<int>::iterator i = slice_points.begin(); i!= slice_points.end(); ++i)
+      {
+        sparam->add_slice_point(*i);
+      }
+    for (auto t:tops)
+      lparam->add_top(t);
+    lparam->add_bottom(bottom);
+  }
+
+  void NetLayersCaffeRecurrent::add_permute(caffe::NetParameter *net_params, std::string top, std::string bottom, int naxis, bool train, bool test)
+  {
+    caffe::LayerParameter *lparam = net_params->add_layer();
+    lparam->set_type("Permute");
+    lparam->set_name("permute_T_N_"+bottom);
+    lparam->add_top(top);
+    lparam->add_bottom(bottom);
+    caffe::PermuteParameter *pparam = lparam->mutable_permute_param();
+    pparam->add_order(1);
+    pparam->add_order(0);
+    pparam->add_order(2);
+    if (naxis == 4)
+      pparam->add_order(3);
+    if (train)
+      {
+        caffe::NetStateRule *nsr;
+        nsr = lparam->add_include();
+        nsr->set_phase(caffe::TRAIN);
+      }
+    else if (test)
+      {
+        caffe::NetStateRule *nsr;
+        nsr = lparam->add_include();
+        nsr->set_phase(caffe::TEST);
+      }
+
+  }
+
+  void NetLayersCaffeRecurrent::add_concat(caffe::NetParameter *net_params,
+                                           std::string name,
+                                           std::string top,
+                                           std::vector<std::string> bottoms,
+                                           int axis)
+  {
+    caffe::LayerParameter *lparam = net_params->add_layer();
+    lparam->set_type("Concat");
+    lparam->set_name(name);
+    lparam->add_top(top);
+    for (auto i : bottoms)
+      {
+        lparam->add_bottom(i);
+      }
+    caffe::ConcatParameter *cparam = lparam->mutable_concat_param();
+    cparam->set_axis(axis);
+  }
+
+  void NetLayersCaffeRecurrent::add_affine(caffe::NetParameter *net_params,
+                                           std::string name,
+                                           std::string bottom,
+                                           std::string top,
+                                           int nout)
+  {
+    caffe::LayerParameter *lparam = net_params->add_layer();
+    lparam->set_type("InnerProduct");
+    lparam->set_name(name);
+    lparam->add_top(top);
+    lparam->add_bottom(bottom);
+    caffe::InnerProductParameter *cparam = lparam->mutable_inner_product_param();
+    cparam->set_num_output(nout);
+    cparam->set_axis(2);
+    caffe::FillerParameter *wfparam = cparam->mutable_weight_filler();
+    wfparam->set_type("xavier");
+    caffe::FillerParameter *bfparam = cparam->mutable_bias_filler();
+    bfparam->set_type("xavier");
+  }
+
+
+  void NetLayersCaffeRecurrent::configure_net(const APIData &ad_mllib)
+  {
+
+    std::vector<std::string> layers = {"L","L"}; // default
+    std::vector<double> dropouts; // default
```

Comment:
I suggest we keep using the existing `dropout` parameter and have the code detect whether this is an array or not. You may need to first check for singleton with `try` and `catch`, then array.

Reply:
done